Enable conditional pre-start-date access to courses

andy-armstrong opened the pull request:
## [LEARNER-1854](https://openedx.atlassian.net/browse/LEARNER-1854)

### Description

This change conditionally allows learners to access courses before their start date (but after their enrollment date). It uses Waffle so that we can enable this only for certain courses so we can test the feasibility of a wider rollout.

### Sandbox

- [x] https://andy-armstrong.sandbox.edx.org/dashboard

I've set up this sandbox so that there are two courses that start in the future: "Future Course" which has been Waffle enabled, and "Future Course 2" which has not. You can try it with verified@example.com which has enrolled in both courses.

### Testing
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
FYI: @robrap @dianakhuang @HarryRein @marcotuts @lizcohen @catong 
 
### Post-review
- [ ] Rebase and squash commits

andy-armstrong commented:
@robrap - I met with @marcotuts and he explained that the information shown on the Progress page comes from the grading policy, not the course content. We agreed that this didn't seem problematic, so I've removed all the logic to try to hide the Progress tab. It would be better to eventually have a message saying that the course hasn't started yet, but that can come later if this experiment proves useful.

andy-armstrong commented:
jenkins run lettuce

andy-armstrong commented:
jenkins run python

andy-armstrong commented:
jenkins run python

andy-armstrong commented:
jenkins run bokchoy

marcotuts commented:
This is likely not related to this sandbox, but I'm not currently seeing dates on the /info page. That's probably ok since this page is going away but I thought I'd mention it in case it is due to other / larger issues.

marcotuts commented:
I'm also not able to see the masquerade bar anymore even though I am a staff member. This bar is necessary / helpful to test the student vs staff visibility rules so I think it blocks further sandbox testing / review.

dianakhuang commented:
@marcotuts the dates should now be showing (configuration flag wasn't turned on for the sandbox).

The question about the masquerade bar showing is making me wonder, because we used to turn that off because it was possible to get into a 'bad' state. With these changes, it might not be an issue, though it will make the code uglier as we handle more cases.

andy-armstrong commented:
@dianakhuang @marcotuts I've improved my fix so that the course access logic and the hide preview logic share the same code to determine whether the course is visible given the start date and the state of my new flag. I've added a unit test, and I've fixed all the query count failures by blacklisting the Waffle queries as @robrap had done in other stories.

I'm rebuilding my sandbox now, and this is ready for re-review.

robrap commented:
jenkins run bokchoy

Bokchoy failure details:
- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/39885/
-
 common.test.acceptance.tests.lms.test_lms_courseware.ProctoredExamTest.test_timed_exam_flow_1_True

Not sure if both ddt tests are failing, so we should split before we mark as flaky (or delete). :)

andy-armstrong commented:
@robrap there seems to be a problem with the course override logic. I tried adding a second course, one that is to be overridden, and one that isn't. Unfortunately they both return False now, and the course that was working with an override has stopped working. My theory is that the problem is that the first value gets cached, and then the second one doesn't even get a chance to get into the code.  I'm going to try to capture the problem in a unit test. Unfortunately it makes my sandbox useless for testing right now!

andy-armstrong commented:
@robrap that test is already reported as flaky: https://openedx.atlassian.net/browse/EDUCATOR-312

andy-armstrong commented:
@robrap It is indeed the caching. I commented out the line that updates the cache, and then the course overrides started working as expected. I think for now I'm going to disable caching if a callback is provided, because the result can differ between calls for the same flag.

andy-armstrong commented:
@marcotuts I've been playing around with options for courses that haven't started:

1. The "Discussion" and "Wiki" tabs work as expected.
2. I tried creating a section with a start date earlier than the course start date, but they still didn't show up. My theory is that the course blocks API is merging the dates and so doesn't allow any date to be earlier than course start. Should this work? If not, how would a course team create an open week 1? Is their only option to set the start date to the enrollment date, but then use dates on sections to keep those closed?
3. I tried creating a welcome message but for some reason it isn't being shown. There may be a bug here, as it is intended to only hide the message for anonymous and unenrolled users.

Is there anything else that should work before a course starts?

andy-armstrong commented on openedx/core/djangoapps/waffle_utils/__init__.py:
```diff
@@ -223,7 +223,9 @@ def is_flag_active(self, flag_name, check_before_waffle_callback=None, flag_unde
                     value = flag_is_active(get_request(), namespaced_flag_name)
 
-            self._cached_flags[namespaced_flag_name] = value
+            # Don't cache the result if it might have been affected by the callback
```
@robrap this is the line of code that is necessary to allow course Waffle flags to work with different values. Let's discuss tomorrow.

robrap replied:
Good call @andy-armstrong.  Thanks.  A couple of additional comments on some possible clean-up.

1. I'd change this comment to:
```
# If wanted, caching for the callback must be done by the callback provider, because the
# cache key may require additional name-spacing.  See CourseWaffleFlag.
```
2. Since I wrote that comment, we should add caching to CourseWaffleFlag for fun. :)  The following is untested sample code:
```
cache_key = '{}.{}'.format(self.namespaced_flag_name, course_key.to_deprecated_string())
value = self. waffle_namespace._cached_flags[cache_key]
if value is None:
   ...
   self. waffle_namespace._cached_flags[cache_key] = value
return value
```
3. You missed some renaming of course_id to course_key.
4. Other possible Waffle Flag cleanup is that strictly speaking, the callback doesn't need to send the flag name because WaffleFlag has a namespaced_flag_name property.  If we don't use this property here, should it be deleted if it has no uses?

andy-armstrong replied:
Thanks @robrap. I realized while walking my dog that this approach doesn't work. Here are some further thoughts:
1. Not caching if the callback is specified means the code essentially never caches any more, since we always use CourseWaffleFlag.
2. The reason we haven't seen the problem before is the problem occurs if the same flag is queried with multiple keys in one request. For most pages, all requests will be for the same course. The dashboard is the first place where we've had two courses queried within a single request.
3. The important thing is that we minimize the number of database queries. The behavior we need is that we only look up each record once, so I think we need separate caches for the Waffle flag and for the override table.

I'm thinking of removing this commit from this PR so that we can merge it. I'll make a JIRA ticket for the bug which we should fix before we start to use this flag in production. We need to write a test where we query across multiple courses within a test. We should probably also write a query verification test that we can make many calls across multiple courses and only have a minimal number of queries made.

robrap replied:
@andy-armstrong Were you responding to me, or were your writing at the same time?

robrap replied:
Also, @andy-armstrong, another good point.  The answer is not to never cache when using the callback, but to [indent this line](https://github.com/edx/edx-platform/blob/a815003b3b96f6cde666c9c4b1d16751db3e19f6/openedx/core/djangoapps/waffle_utils/__init__.py#L226-L226) so that we don't cache if the value we are getting is from the callback.

robrap replied:
@andy-armstrong: I've got my fingers crossed that I can land my story. Let me know if you need/want me to work on these waffle changes in the morning.

andy-armstrong commented:
@marcotuts @dianakhuang @robrap I removed the commit to fix the course overrides for Waffle flags, so that will be handled in a subsequent PR. I left that commit on the sandbox so you can still see the behavior. I think this code is ready to merge as is, and if we want to do any improvements (e.g. to show welcome messages) we should do them in follow-on stories. @dianakhuang, once you've approved the current set of changes I will merge this PR.

andy-armstrong commented on openedx/features/course_experience/utils.py:
```diff
@@ -71,9 +71,9 @@ def mark_last_accessed(user, course_key, block):
-    set_last_accessed_default(course_outline_root_block)
-    mark_last_accessed(request.user, course_key, course_outline_root_block)
-
+    course_outline_root_block = all_blocks['blocks'].get(all_blocks['root'], None)
```
Exactly! This was crashing when I first added the flag.

andy-armstrong commented:
jenkins run bokchoy
